public/recorder.js

```diff
@@ -41,9 +56,186 @@ coverImg.appendChild(avatar);
 coverTextInp.value = randomEmoji;
 
 
+function resetAudioGraph() {
+  if (playbackNode) {
+    playbackNode.disconnect();
+    playbackNode.removeEventListener("ended", onEndedHandler);
+    playbackNode = null;
+  }
+  if (refNode) {
+    refNode.disconnect();
+    refNode.removeEventListener("ended", onEndedHandler);
+    refNode = null;
+  }
+}
+
+function onEndedHandler() {
+  audioContext.suspend();
+  resetAudioGraph();
+  stopKaraoke();
+
+  if (recorder) {
+    recorder.stop();
+    stopLabel.textContent = "Stop";
+  }
+
+  recordBtn.disabled = false;
+  playBtn.disabled = false;
+  pauseBtn.disabled = true;
+  stopBtn.disabled = true;
+}
+
+
+async function prepareReferencePlayback() {
+  startTime = audioContext.currentTime;
+  playbackNode = audioContext.createBufferSource();
+  playbackNode.buffer = refDecodedAudio;
+  playbackNode.start(startTime);
+  playbackNode.connect(audioContext.destination);
+  playbackNode.addEventListener("ended", onEndedHandler);
+}
+
+async function prepareCountdownAndReferencePlayback() {
+  startTime = audioContext.currentTime;
+  playbackNode = audioContext.createGain();
+
+  const toneNode1 = audioContext.createBufferSource();
+  toneNode1.buffer = toneDecodedAudio;
+  toneNode1.start(startTime);
+  toneNode1.stop(startTime + 1);
+  toneNode1.connect(playbackNode);
+
+  const toneNode2 = audioContext.createBufferSource();
+  toneNode2.buffer = toneDecodedAudio;
+  toneNode2.start(startTime + 1);
+  toneNode2.stop(startTime + 2);
+  toneNode2.connect(playbackNode);
+
+  const toneNode3 = audioContext.createBufferSource();
+  toneNode3.buffer = toneDecodedAudio;
+  toneNode3.start(startTime + 2);
+  toneNode3.stop(startTime + 3);
+  toneNode3.connect(playbackNode);
+
+  refNode = audioContext.createBufferSource();
+  refNode.buffer = refDecodedAudio;
+  refNode.start(startTime + 3);
+
+  refNode.connect(playbackNode);
+  playbackNode.connect(audioContext.destination);
+
+  refNode.addEventListener("ended", onEndedHandler);
+}
+
+async function prepareRecordingPlayback() {
+  // Safari compat notes:
+  // - No support for blob.arrayBuffer() yet, so need to use FileReader
+  // - No support for Promise version of decodeAudioData yet
+  const buffer = await new Promise(resolve => {
+    const reader = new FileReader();
+    reader.readAsArrayBuffer(recording);
+    reader.onloadend = () => {
+      resolve(reader.result);
+    }
+  });
+  const decodedAudio = await new Promise((res, rej) => audioContext.decodeAudioData(buffer, res, rej));
+  startTime = audioContext.currentTime;
+  playbackNode = audioContext.createBufferSource();
+  playbackNode.buffer = decodedAudio;
+  playbackNode.start(startTime);
+  playbackNode.connect(audioContext.destination);
+  playbackNode.addEventListener("ended", onEndedHandler);
+}
+
+function startKaraoke({ wait } = { wait: 0 }) {
+  let lastSegment = -1;
+  let lastLang = null;
+
+  function scheduleNextCheck() {
+    karaokeTimeout = setTimeout(karaokeCheck, 100);
```

Comment:
is the intent here to have something less frequent than ontimeupdate?
Was 100ms chosen for a particular reason? would rAF be too aggressive still?

Reply:
Actually, `timeupdate` events tend to fire every 250ms depending on browsers (spec says every 15ms to 250ms), so the goal was the opposite as I noticed that some of the updates happened slightly too late. We can certainly lower but that does not seem needed and rAF indeed seems a bit aggressive for that.